python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py

```diff
@@ -388,10 +388,20 @@ def __init__(
         self._start_timestamp = context.create_time.timestamp()
         self._end_timestamp = get_current_timestamp()
 
+        # hidden_param which should only be set by legacy RuleConditions
         self._subsets_with_metadata = check.opt_sequence_param(
-            subsets_with_metadata, "subsets_with_metadata", AssetSubsetWithMetadata
+            kwargs.get("subsets_with_metadata"), "subsets_with_metadata", AssetSubsetWithMetadata
         )
-        self._extra_state = extra_state
+
+        # hidden_param which should only be set by builtin conditions which require high performance
+        # in their serdes layer
+        extra_state = kwargs.get("extra_state")
+        check.param_invariant(
+            not (cursor and extra_state),
+            "extra_state",
+            "Cannot provide both cursor and extra_state.",
+        )
+        self._extra_state = check.opt_str_param(cursor, "cursor") or extra_state
```

Comment:
Well I think we should offer the same interface to users as well.

What if we instead typed `ExtraState` as `StructuredCursor` then typed

```
cursor: Union[str, StructuredCursor]
```

I think that would make intent more clear and avoid an invalid state that is only caught at runtime.

Reply:
@schrockn The issue with that is that the read side gets more complicated as the user (who almost exclusively will just be writing a string value to the cursor) will now have to handle a type that's potentially not a string when they access `context.cursor`.

My feeling is that a bare string cursor has been sufficient for sensors, and so it seems likely that it will be sufficient for AutomationConditions as well. I updated the name of the hidden parameter to `structured_cursor` to make it more obvious that these parameters intersect with each other. My feeling is that if in the future we see demand for higher-performance cursors in user code, we can un-hide this parameter.